_posts/2019-12-25-technical-debt.md

```diff
@@ -0,0 +1,107 @@
+
+<br/>
+> Debt . . . . that peculiar nexus where money, narrative or story, and religious belief intersect, often with explosive force. - Margaret Atwood
+
+<br/>
+Technical debt is an infamous subject in software development. The common understanding is that we accrue it by some series of faustian pacts; short-cuts are taken in order to deliver a feature sooner, with the underlying assumption that additional features will take longer to complete. A development team that continues to accrue technical debt will eventually find time taken to deliver the next feature will begin to approach infinity. In defiance of what is an obvious problem to the average developer, business stakeholders are often loathe to address it. Technical debt is unyielding to measure and resistant to containment due to its nature as an abstract concept. If developers wish relief from technical debt, a better way must be found to describe the cost in real terms.
+
+There is a long list of potential consequences for allowing the unrestrained accrual of technical debt. A reduction in the ability to deliver features (usually referred to as 'velocity') is chief among them. Knowing the consequences of even the smallest change to the system is becomes difficult, with minor changes provide an opportunity to observe the butterfly effect. The engineer that coined the term technical debt, Ward Cunningham (*), explains that the core of technical debt is understanding. This lack of understanding has prevents engineers from safely modifying a system, and handicaps our ability to prevent outages. When those outages eventually occur (and they will), the knowledge deficit will frustrate any attempts at triage.
+
+<br/>
+> i have a theory, which is that we struggle to get the time allocated to pay down technical debt (or improve deploys, etc) because to biz types we basically sound like the underpants gnomes. step 1 ... pay down technical debt. step 2 ... step 3 .. PROFIT - Charity Majors, CTO @ Honeycomb.io
+
+<br/>
+Persuading stakeholders to address technical debt is fraught with difficulty. Requesting to pay down technical debt is akin to asking someone to trade something they confidently reason has some value, for a loose promise that if and when they ask for something else later, it may be easier to deliver. It is little wonder that most technical debt doomsaying fails to convince anyone, notably in the low-trust terrain that forms the modern enterprise. The benefits of paying down technical debt are often presented in a nebulous way, which is in contrast to the concrete nature of getting a feature out to market sooner. Technical debt may be a murky concept, but that does not mean that the benefits of paying it down are. If we want technical debt to be taken in earnest, we need to get better at gauging the effects and presenting a case for removal that has discernable benefit.
```

Comment:
There's an assumption here that the development team requires permission to pay down technical debt.  And talks to responsibility/accountability of the software development team

Asking for permission is akin to throwing code over the wall - it provides cognitive dissonance between the introduction and it's repayment.  

The Software Craftsmanship movement was all about taking back control, and responsibility over a teams technical decision-making.

Reply:
I get that and I agree; but that's the reality that most software development teams are living with.

Most of us are just playing with somebody elses money.